File: alchemist-incarnation-protelis/src/main/java/it/unibo/alchemist/model/implementations/actions/RunProtelisProgram.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.implementations.actions;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import it.unibo.alchemist.model.implementations.molecules.SimpleMolecule;
import it.unibo.alchemist.model.interfaces.Action;
import it.unibo.alchemist.model.interfaces.Context;
import it.unibo.alchemist.model.interfaces.Dependency;
import it.unibo.alchemist.model.interfaces.Environment;
import it.unibo.alchemist.model.interfaces.Molecule;
import it.unibo.alchemist.model.interfaces.Node;
import it.unibo.alchemist.model.interfaces.Position;
import it.unibo.alchemist.model.interfaces.Reaction;
import it.unibo.alchemist.model.interfaces.capabilities.ProtelisCapability;
import it.unibo.alchemist.protelis.AlchemistExecutionContext;
import it.unibo.alchemist.protelis.AlchemistNetworkManager;
import org.apache.commons.math3.distribution.RealDistribution;
import org.apache.commons.math3.random.RandomGenerator;
import org.danilopianini.util.ImmutableListSet;
import org.danilopianini.util.ListSet;
import org.protelis.lang.ProtelisLoader;
import org.protelis.vm.ProtelisProgram;
import org.protelis.vm.ProtelisVM;

import java.io.IOException;
import java.io.ObjectInputStream;

import static it.unibo.alchemist.model.math.RealDistributionUtil.makeRealDistribution;
import static java.util.Objects.requireNonNull;

/**
 * @param <P> position type
 */
public final class RunProtelisProgram<P extends Position<P>> implements Action<Object> {

    private static final long serialVersionUID = 1L;
    private static final String
        EI_EXPOSE_REP = "EI_EXPOSE_REP",
        INTENTIONAL = "This is intentional";
    private boolean computationalCycleComplete;
    private final Environment<Object, P> environment;
    private final Molecule name;
    private final Node<Object> node;
    private String originalProgram = "unknown";
    private final ProtelisProgram program;
    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "All the random engines provided by Apache are Serializable")
    private final RandomGenerator random;
    private final Reaction<Object> reaction;
    private final double retentionTime;
    private final AlchemistNetworkManager networkManager;
    private transient ProtelisVM vm;
    private transient AlchemistExecutionContext<P> executionContext;

    private RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final ProtelisProgram program,
            final double retentionTime,
            final RealDistribution packetLossDistance
    ) {
        final var otherCopies = node.getReactions().stream()
            .flatMap(it -> it.getActions().stream())
            .filter(it -> it instanceof RunProtelisProgram)
            .map(it -> ((RunProtelisProgram) it).program.getName())
            .filter(programName -> programName.equals(program.getName()))
            .count();
        name = new SimpleMolecule(program.getName() + (otherCopies == 0 ? "" : "$copy" + otherCopies));
        this.program = requireNonNull(program);
        this.environment = requireNonNull(environment);
        this.node = requireNonNull(node);
        random = requireNonNull(randomGenerator);
        this.reaction = requireNonNull(reaction);
        networkManager = new AlchemistNetworkManager(reaction, this, retentionTime, packetLossDistance);
        this.node.asCapability(ProtelisCapability.class).addNetworkManger(this, networkManager);
        executionContext = new AlchemistExecutionContext<>(environment, node, reaction, randomGenerator, networkManager);
        vm = new ProtelisVM(program, executionContext);
        this.retentionTime = retentionTime;
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final String program
    ) {
        this(environment, node, reaction, randomGenerator, program, Double.NaN);
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @param retentionTime
     *            how long the messages will be stored. Pass {@link Double#NaN}
     *            to mean that they should get eliminated upon node awake.
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final String program,
            final double retentionTime
    ) {
        this(environment, node, reaction, randomGenerator, program, retentionTime, null);
        originalProgram = program;
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @param packetLossDistributionName
     *            the package loss probability, scaling with distance.
     *            This is the name of the {@link RealDistribution} to be used as follows:
     *            its PDF will be computed with {@link RealDistribution#density(double)},
     *            and will be fed the distance between the current node and the neighbor;
     *            the generated probability will in turn be used to determine the probability of the package to be
     *            successfully delivered.
     * @param packetLossDistributionParameters
     *            parameters that will be passed when building the packet loss distribution
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final String program,
            final String packetLossDistributionName,
            final double... packetLossDistributionParameters
    ) {
        this(
            environment,
            node,
            reaction,
            randomGenerator,
            program,
            Double.NaN,
            makeRealDistribution(randomGenerator, packetLossDistributionName, packetLossDistributionParameters)
        );
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @param retentionTime
     *            how long the messages will be stored. Pass {@link Double#NaN}
     *            to mean that they should get eliminated upon node awake.
     * @param packetLossDistributionName
     *            the package loss probability, scaling with distance.
     *            This is the name of the {@link RealDistribution} to be used as follows:
     *            its PDF will be computed with {@link RealDistribution#density(double)},
     *            and will be fed the distance between the current node and the neighbor;
     *            the generated probability will in turn be used to determine the probability of the package to be
     *            successfully delivered.
     * @param packetLossDistributionParameters
     *            parameters that will be passed when building the packet loss distribution
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final String program,
            final double retentionTime,
            final String packetLossDistributionName,
            final double... packetLossDistributionParameters
    ) {
        this(
            environment,
            node,
            reaction,
            randomGenerator,
            program,
            retentionTime,
            makeRealDistribution(randomGenerator, packetLossDistributionName, packetLossDistributionParameters)
        );
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @param retentionTime
     *            how long the messages will be stored. Pass {@link Double#NaN}
     *            to mean that they should get eliminated upon node awake.
     * @param packetLossDistributionName
     *            the package loss probability, scaling with distance.
     *            This is the name of the {@link RealDistribution} to be used as follows:
     *            its PDF will be computed with {@link RealDistribution#density(double)},
     *            and will be fed the distance between the current node and the neighbor;
     *            the generated probability will in turn be used to determine the probability of the package to be
     *            successfully delivered.
     * @param packetLossDistributionParameters
     *            parameters that will be passed when building the packet loss distribution
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final ProtelisProgram program,
            final double retentionTime,
            final String packetLossDistributionName,
            final double... packetLossDistributionParameters
    ) {
        this(
            environment,
            node,
            reaction,
            randomGenerator,
            program,
            retentionTime,
            makeRealDistribution(randomGenerator, packetLossDistributionName, packetLossDistributionParameters)
        );
    }

    /**
     * @param environment
     *            the environment
     * @param node
     *            the node
     * @param reaction
     *            the reaction
     * @param randomGenerator
     *            the random engine
     * @param program
     *            the Protelis program
     * @param retentionTime
     *            how long the messages will be stored. Pass {@link Double#NaN}
     *            to mean that they should get eliminated upon node awake.
     * @param packetLossDistance
     *            the package loss probability, scaling with distance.
     *            This {@link RealDistribution} will be used as follows:
     *            its PDF will be computed with {@link RealDistribution#density(double)},
     *            and will be fed the distance between the current node and the neighbor;
     *            the generated probability will in turn be used to determine the probability of the package to be
     *            successfully delivered. Can be null, in which case packets always arrive to neighbors.
     * @throws SecurityException
     *             if you are not authorized to load required classes
     */
    public RunProtelisProgram(
            final Environment<Object, P> environment,
            final Node<Object> node,
            final Reaction<Object> reaction,
            final RandomGenerator randomGenerator,
            final String program,
            final double retentionTime,
            final RealDistribution packetLossDistance
    ) {
        this(
            environment,
            node,
            reaction,
            randomGenerator,
            ProtelisLoader.parse(program),
            retentionTime,
            packetLossDistance
        );
        originalProgram = program;
    }

    /**
     * @return the molecule associated with the execution of this program
     */
    public Molecule asMolecule() {
        return name;
    }

    @Override
    @SuppressWarnings("unchecked")
    public RunProtelisProgram<P> cloneAction(final Node<Object> node, final Reaction<Object> reaction) {
        if (node.asCapabilityOrNull(ProtelisCapability.class) != null) {
            try {
                return new RunProtelisProgram<P>(
                    getEnvironment(),
                    node,
                    reaction,
                    getRandomGenerator(),
                    originalProgram,
                    getRetentionTime()
                );
            } catch (SecurityException e) {
                throw new IllegalStateException(e);
            }
        }
        throw new IllegalArgumentException("The node must have an instance of " + ProtelisCapability.class.getSimpleName());
    }

    @Override
    public boolean equals(final Object other) {
        if (other == this) {
            return true;
        }
        if (other != null && other.getClass() == getClass()) {
            final var otherProgram = (RunProtelisProgram<?>) other;
            return name.equals(otherProgram.name);
        }
        return false;
    }

    @Override
    public void execute() {
        vm.runCycle();
        node.setConcentration(name, vm.getCurrentValue());
        computationalCycleComplete = true;
    }

    @Override
    public Context getContext() {
        /*
         * A Protelis program never writes in other nodes
         */
        return Context.LOCAL;
    }

    /**
     * @return the environment
     */
    @SuppressFBWarnings(value = EI_EXPOSE_REP, justification = INTENTIONAL)
    public Environment<Object, P> getEnvironment() {
        return environment;
    }

    @Override
    public ListSet<? extends Dependency> getOutboundDependencies() {
        return ImmutableListSet.of(Dependency.EVERY_MOLECULE);
    }

    /**
     * @return the node
     */
    @SuppressFBWarnings(value = EI_EXPOSE_REP, justification = INTENTIONAL)
    public Node<Object> getNode() {
        return node;
    }

    /**
     * @return the internal {@link RandomGenerator}
     */
    @SuppressFBWarnings(value = EI_EXPOSE_REP, justification = INTENTIONAL)
    public RandomGenerator getRandomGenerator() {
        return random;
    }

    /**
     * @return the retention time
     */
    public double getRetentionTime() {
        return retentionTime;
    }

    @Override
    public int hashCode() {
        return name.hashCode();
    }

    /**
     * @return true if the Program has finished its last computation,
     * and is ready to send a new message (used for dependency management)
     */
    public boolean isComputationalCycleComplete() {
        return computationalCycleComplete;
    }

    /**
     * Resets the computation status (used for dependency management).
     */
    public void prepareForComputationalCycle() {
        this.computationalCycleComplete = false;
    }

    private void readObject(final ObjectInputStream stream) throws ClassNotFoundException, IOException {
        stream.defaultReadObject();
        executionContext = new AlchemistExecutionContext<>(environment, node, reaction, random, networkManager);
        vm = new ProtelisVM(program, executionContext);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toString() {
        return name + "@" + node.getId();
    }

    /**
     * Provides an access to the underlying {@link org.protelis.vm.ExecutionContext}.
     *
     * @return the current {@link AlchemistExecutionContext}
     */
    @SuppressFBWarnings(value = EI_EXPOSE_REP, justification = INTENTIONAL)
    public AlchemistExecutionContext<P> getExecutionContext() {
        return executionContext;
    }
}

File: alchemist-incarnation-protelis/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/RunProtelisProgram.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.actions

import it.unibo.alchemist.model.implementations.molecules.SimpleMolecule
import it.unibo.alchemist.model.implementations.properties.ProtelisDevice
import it.unibo.alchemist.model.interfaces.Action
import it.unibo.alchemist.model.interfaces.Context
import it.unibo.alchemist.model.interfaces.Dependency
import it.unibo.alchemist.model.interfaces.Environment
import it.unibo.alchemist.model.interfaces.Molecule
import it.unibo.alchemist.model.interfaces.Node
import it.unibo.alchemist.model.interfaces.Node.Companion.asProperty
import it.unibo.alchemist.model.interfaces.Position
import it.unibo.alchemist.model.interfaces.Reaction
import it.unibo.alchemist.model.math.RealDistributionUtil
import it.unibo.alchemist.protelis.AlchemistExecutionContext
import it.unibo.alchemist.protelis.AlchemistNetworkManager
import org.apache.commons.math3.distribution.RealDistribution
import org.apache.commons.math3.random.RandomGenerator
import org.danilopianini.util.ImmutableListSet
import org.protelis.lang.ProtelisLoader
import org.protelis.vm.ProtelisProgram
import org.protelis.vm.ProtelisVM
import java.io.ObjectInputStream

/**
 * An [Action] that executes a Protelis program.
 *
 * Requires the current [randomGenerator] and [environment], a valid [ProtelisDevice] ([device]),
 * and the local [reaction] hosting the computation.
 *
 * The program can be created using a String ([originalProgram]), or, alternatively,
 * by providing a [ProtelisProgram] ([program]).
 *
 * [retentionTime] specifies whether, upon message usage, the received messages should be deleted
 * (assuming a reasonable synchronization among devices) or if they should remain in memory for a specified amount
 * of time. By default, [retentionTime] is [Double.NaN], indicating that messages are deleted upon read.
 *
 * It is possible to sumulate the loss of messages due to a higher connection distance by providing a [RealDistribution]
 * ([packetLossDistance]) mapping distances to the loss probability. By default this feature is disabled.
 */
class RunProtelisProgram<P : Position<P>> private constructor(
    val randomGenerator: RandomGenerator,
    val environment: Environment<Any, P>,
    val device: ProtelisDevice<P>,
    val reaction: Reaction<Any>,
    val originalProgram: String,
    val program: ProtelisProgram,
    val retentionTime: Double,
    val packetLossDistance: RealDistribution?,
) : Action<Any> {

    @JvmOverloads constructor(
        randomGenerator: RandomGenerator,
        environment: Environment<Any, P>,
        device: ProtelisDevice<P>,
        reaction: Reaction<Any>,
        program: ProtelisProgram,
        retentionTime: Double = Double.NaN,
    ) : this(
        randomGenerator,
        environment,
        device,
        reaction,
        originalProgram = program.name,
        program = program,
        retentionTime = retentionTime,
        packetLossDistance = null
    )

    @JvmOverloads constructor(
        randomGenerator: RandomGenerator,
        environment: Environment<Any, P>,
        device: ProtelisDevice<P>,
        reaction: Reaction<Any>,
        program: ProtelisProgram,
        retentionTime: Double = Double.NaN,
        packetLossDistributionName: String,
        vararg packetLossDistributionParameters: Double,
    ) : this(
        randomGenerator,
        environment,
        device,
        reaction,
        originalProgram = program.name,
        program = program,
        retentionTime = retentionTime,
        packetLossDistance = RealDistributionUtil.makeRealDistribution(
            randomGenerator,
            packetLossDistributionName,
            *packetLossDistributionParameters
        )
    )

    @JvmOverloads constructor(
        randomGenerator: RandomGenerator,
        environment: Environment<Any, P>,
        device: ProtelisDevice<P>,
        reaction: Reaction<Any>,
        program: String,
        retentionTime: Double = Double.NaN
    ) : this(
        randomGenerator,
        environment,
        device,
        reaction,
        originalProgram = program,
        program = ProtelisLoader.parse(program),
        retentionTime = retentionTime,
        packetLossDistance = null,
    )

    @JvmOverloads constructor(
        randomGenerator: RandomGenerator,
        environment: Environment<Any, P>,
        device: ProtelisDevice<P>,
        reaction: Reaction<Any>,
        program: String,
        retentionTime: Double = Double.NaN,
        packetLossDistributionName: String,
        vararg packetLossDistributionParameters: Double,
    ) : this(
        randomGenerator,
        environment,
        device,
        reaction,
        originalProgram = program,
        retentionTime = retentionTime,
        program = ProtelisLoader.parse(program),
        packetLossDistance = RealDistributionUtil.makeRealDistribution(
            randomGenerator,
            packetLossDistributionName,
            *packetLossDistributionParameters
        )
    )

    /**
     * The Alchemist [Node] hosting the [ProtelisDevice].
     */
    val node = device.node

    /**
     * @return true if the Program has finished its last computation,
     * and is ready to send a new message (used for dependency management)
     */
    var isComputationalCycleComplete = false
        private set

    private val name: Molecule = node.reactions.asSequence()
        .flatMap { it.actions.asSequence() }
        .filterIsInstance<RunProtelisProgram<*>>()
        .map { it.program.name }
        .count { it == program.name }
        .let { otherCopies -> SimpleMolecule(program.name + if (otherCopies == 0) "" else "\$copy$otherCopies") }

    private val networkManager = AlchemistNetworkManager(reaction, device, this, retentionTime, packetLossDistance)

    /**
     * Provides an access to the underlying [org.protelis.vm.ExecutionContext].
     *
     * @return the current [AlchemistExecutionContext]
     */
    @Transient
    var executionContext = AlchemistExecutionContext(environment, node, reaction, randomGenerator, networkManager)
        private set

    @Transient
    private var vm: ProtelisVM = ProtelisVM(program, executionContext)

    init {
        device.addNetworkManger(this, networkManager)
    }

    /**
     * @return the molecule associated with the execution of this program
     */
    fun asMolecule(): Molecule {
        return name
    }

    override fun cloneAction(node: Node<Any>, reaction: Reaction<Any>): RunProtelisProgram<P> = RunProtelisProgram(
        randomGenerator,
        environment,
        node.asProperty(),
        reaction,
        originalProgram = originalProgram,
        program = program, // TODO: this is broken until https://github.com/Protelis/Protelis/pull/676 gets merged
        retentionTime = retentionTime,
        packetLossDistance = packetLossDistance,
    )

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        if (other != null && other.javaClass == javaClass) {
            val otherProgram = other as RunProtelisProgram<*>
            return name == otherProgram.name
        }
        return false
    }

    override fun execute() {
        vm.runCycle()
        node.setConcentration(name, vm.currentValue)
        isComputationalCycleComplete = true
    }

    /*
     * A Protelis program never writes in other nodes
     */
    override fun getContext() = Context.LOCAL

    override fun getOutboundDependencies(): ImmutableListSet<Dependency> =
        ImmutableListSet.of(Dependency.EVERY_MOLECULE)

    override fun hashCode() = name.hashCode()

    /**
     * Resets the computation status (used for dependency management).
     */
    fun prepareForComputationalCycle() {
        isComputationalCycleComplete = false
    }

    @Suppress("UnusedPrivateMember")
    private fun readObject(stream: ObjectInputStream) {
        stream.defaultReadObject()
        executionContext = AlchemistExecutionContext(environment, node, reaction, randomGenerator, networkManager)
        vm = ProtelisVM(program, executionContext)
    }

    override fun toString(): String {
        return name.toString() + "@" + node.id
    }

    companion object {
        private const val serialVersionUID = 2L
    }
}
